go/roothash/api/message/incoming_message.go

```diff
@@ -26,10 +29,29 @@ type IncomingMessage struct {
 	// transferred before the message is processed by the runtime.
 	Tokens quantity.Quantity `json:"tokens,omitempty"`
 
-	// Data is arbitrary runtime-dependent data.
+	// Data is a serialized IncomingMessageData.
 	Data []byte `json:"data,omitempty"`
 }
 
+const LatestIncomingMessageDataVersion = 1
```

Comment:
Why does Core require knowledge of the incoming message format? It should be completely up to the runtime? All incoming messages should be processed/propagated in the same way?

Reply:
for executor workers, they must have all incoming message transactions ready to populate the proposed batch, all before entering the runtime for execution. in this design, we extract the transaction within core

Reply:
hm maybe we don't even have to bump the runtime host protocol version. in the current design we're not even adding a new entry into the runtime to parse out the transactions, because we're exposing it to the node via this struct.

we can add a slot for completely opaque data next to `tx`, if that's desired. or we can add it later when runtimes want to start using such data

Reply:
Hm, am I missing something? All incoming messages should be handled in the same way, regardless of what is in them or how the runtime interprets them. E.g. they would all be handled as "transactions", not only some subset of such messages. Otherwise this will make things more complicated.

Reply:
ya let's make that simplification, that way there's no parsing to be done, and thus no risk of different parsing

Reply:
thusly simplified